core/unix/signal.c

```diff
@@ -8607,8 +8778,10 @@ handle_suspend_signal(dcontext_t *dcontext, kernel_siginfo_t *siginfo,
     }
     LOG(THREAD, LOG_ASYNCH, 2, "handle_suspend_signal: awake now\n");
 
-    /* re-block so our exit from main_signal_handler is not interrupted */
-    sigprocmask_syscall(SIG_SETMASK, &prevmask, NULL, sizeof(prevmask));
+    if (unblocked_sigs) {
+        /* re-block so our exit from main_signal_handler is not interrupted */
+        sigprocmask_syscall(SIG_SETMASK, &prevmask, NULL, sizeof(prevmask));
+    }
```

Comment:
I created api.detach_signal to test the native signal delivery path but I see no mention of it (or any automated test) in the PR description.  Please evaluate whether api.detach_signal exercises the various paths of this new code and tweak it to reproduce in at least some runs if not.

Reply:
Isn't that just because the test has large stack sizes?  It doesn't need any signal code path changes: just needs a tiny stack to repro that part I would assume.  It is probably worth spending a little time making the stack tiny to reproduce this; if that works maybe leave it tiny or parameterize and run some tiny and some regular.

Reply:
Lowering the altstack size in detach_signal reproduces this. Thanks for the suggestion.

>  if that works maybe leave it tiny or parameterize and run some tiny and some regular.

Half of the threads have the earlier altstack size and others have the smaller one.